Tests/SentryProfilerTests/SentryProfilerTests.mm

```diff
@@ -168,12 +171,118 @@ - (void)testProfilerMutationDuringSlicing
     [self waitForExpectationsWithTimeout:1 handler:nil];
 }
 
+- (void)testProfilerMutationDuringSerialization
```

Comment:
`m`: It looks like we could mutate the profiling data on another thread while serializing. Maybe it makes sense to add a test similar to 
https://github.com/getsentry/sentry-cocoa/blob/1bf8571322c66325de9718c4e6e850f784d22411/Tests/SentryTests/TestLogOutput.swift#L31-L38

Reply:
Can you share the process by which this thread-unsafe mutation can still occur? This PR is making the change to prevent exactly that situation. `_gDataStructureLock` guards the point where those data structures are mutated: https://github.com/getsentry/sentry-cocoa/blob/971805f9f02eef494b69328ccfe7a81f15fc5e28/Sources/Sentry/SentryProfiler.mm#L160-L161

Reply:
and then where those data structures are copied: https://github.com/getsentry/sentry-cocoa/blob/971805f9f02eef494b69328ccfe7a81f15fc5e28/Sources/Sentry/SentryProfiler.mm#L280C39-L285

Reply:
My apologies. I suggested a solution to a problem I didn't explain. When I remove the following line of code https://github.com/getsentry/sentry-cocoa/blob/a337836ce7a68f2b8f9c1a304f6abbf011a44ca0/Sources/Sentry/SentryProfiler.mm#L280

we still get crashes when serializing the data. It would be great if we could develop a test ensuring that we have the proper mechanisms in place so that such crashes don't happen anymore.

Reply:
Thanks for following up Philipp! So this PR is really doing two things:
- doing the copies of the data structures to prevent modification of the data during serialization (the `copy` operations)
- forcing exclusive access to the data structures preventing a race condition (the new and expanded critical sections guarded by the locks)

If you were to convert the calls to `copy` right below that lock guard to simple assignments to the same data structures, then you'd see the new tests I wrote fail.

The locks are only there to prevent the data race on those data structures, which is only going to be a very small edge case that actually causes a crash. The stacktrace in the PR description shows that there's a block enumeration in the call to isValidJSONObject, where if these data structures are modified in `processBacktrace` at just that time, a crash will occur. I think there could even be a data race with the copy calls if there were no locks, because what happens if the data structures are modified from another thread _during_ the execution of the copy?